### DISCUSSION TITLE
difference equation-rhs of ANARX-model

### DISCUSSION THREAD

**Discussion by Duma46** [do-mpc/do-mpc#304] [Q&A]
Good evening everyone :) ,

i trained an artificial neural network with ANARX (Additive Nonlinear AutoregRessive with
eXogenous Inputs) structure  with an external system simulation and evaluated the ANN that simulation (Comparison between ANN prediction and external simulation result). Then i exported the weights of the trained ANN into my do_mpc-project, because i want to use the ANN as do_mpc model. 

According to this paper
https://www.researchgate.net/publication/223002956_On_realizability_of_neural_networks-based_input-output_models_in_the_classical_state-space_form
you can recive the  state-space representation of the ANARX-NN like this.

![image](https://user-images.githubusercontent.com/91143739/192843142-06d3ac63-41b7-43d5-97b0-ef887eaf5512.png)

So i formulated the rhs of discrete model in the template_model of my do_mpc-project, which looks like this with the sigmoid activation-function (1/(1+exp(-x))

def template_model(symvar_type='SX'):
    
    model_type = 'discrete' # either 'discrete' or 'continuous'
    model = do_mpc.model.Model(model_type)
    
    x_1 = model.set_variable(var_type='_x', var_name="x_1")
    u_k = model.set_variable(var_type='_u', var_name="u_k")
    
 
    
    x_2 = model.set_variable(var_type='_x', var_name="x_2")
    x_3 = model.set_variable(var_type='_x', var_name="x_3")
    x_4 = model.set_variable(var_type='_x', var_name="x_4")
    x_5 = model.set_variable(var_type='_x', var_name="x_5")
    
           
    model.set_rhs('x_1',  x_2 + C_storage[0,:,:] @ (1/(1+exp(-(W_storage[0,:,0]*x_1 + W_storage[0,:,1]*u_k ) ))) ) 
    model.set_rhs('x_2',  x_3 + C_storage[1,:,:] @ (1/(1+exp(-(W_storage[1,:,0]*x_1 + W_storage[1,:,1]*u_k) ))) ) 
    model.set_rhs('x_3',  x_4 + C_storage[2,:,:] @ (1/(1+exp(-(W_storage[2,:,0]*x_1 + W_storage[2,:,1]*u_k) ))) ) 
    model.set_rhs('x_4',  x_5 + C_storage[3,:,:] @ (1/(1+exp(-(W_storage[3,:,0]*x_1 + W_storage[3,:,1]*u_k) ))) ) 
    model.set_rhs('x_5',  C_storage[4,:,:] @ (1/(1+exp(-(W_storage[4,:,0]*x_1 + W_storage[4,:,1]*u_k) ))) ) 
                  
    
    model.setup()
    
    return model

The weights C and W have this size
W_storage => ((5,4,2))

C_storage  => ((5,1,4))

**Can someone say if i formulated the rhs of the ODE right?** 
and 
**Is it practicable to use the do_mpc-simulation alone (without mpc or mhe)?**

I am asking, because while evaluating (Comparison between ANARX/do_mpc prediction and external simulation result) the do_mpc model/simulation (ANARX in do_mpc) shows a different behavior (oscillating or very inert) than in the first evaluation (state-space representation in Tensorflow).

I try to figure out the reason of this change in behavior and would be very grateful for your help.  :) 

My simulation setup looks like this:

def template_simulator(model):
    simulator = do_mpc.simulator.Simulator(model)
    
    params_simulator = {
    't_step': 0.005
    }
    
    # Set parameter(s):
    simulator.set_param(**params_simulator)

    # Optional: Set function for parameters and time-varying parameters.
    # Get the template
    tvp_template = simulator.get_tvp_template()
    def tvp_fun(t_now):
        
        return tvp_template
    
    # Set the tvp_fun:
    simulator.set_tvp_fun(tvp_fun)

    # Setup simulator:
    simulator.setup()

    return simulator
